sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/view.scala

```diff
@@ -73,18 +73,23 @@ object EliminateView extends Rule[LogicalPlan] with CastSupport {
       }
       // Map the attributes in the query output to the attributes in the view output by index.
       val newOutput = output.zip(queryOutput).map {
-        case (attr, originAttr) if !attr.semanticEquals(originAttr) =>
+        case (attr, originAttr)
+          if attr.name != originAttr.name || attr.dataType != originAttr.dataType =>
           // `CheckAnalysis` already guarantees that the cast is a up-cast for sure.
           Alias(cast(originAttr, attr.dataType), attr.name)(exprId = attr.exprId,
             qualifier = attr.qualifier, explicitMetadata = Some(attr.metadata))
         case (_, originAttr) => originAttr
       }
-      Project(newOutput, child)
+      if (newOutput != queryOutput) {
+        Project(newOutput, child) -> Nil
```

Comment:
Adding project doesn't mean we add alias for each column. We need to provider attr mapping here as well.

Reply:
I see, because we don't compare exprId any more.